File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& arr, int k) {
        int n = arr.size(); 
        int count = 0; 
        unordered_map<int, int> mp;
        vector<int> pre(n); 

        pre[0] = arr[0];
        for (int i = 1; i < n; i++) {
            pre[i] = pre[i - 1] + arr[i];
        }

        mp[0] = 1;
        for (int i = 0; i < n; i++) {
            int currSum = pre[i];
            int rem = currSum - k;
            if (mp.find(rem) != mp.end()) {
                count += mp[rem];
            }
            mp[currSum]++;
        }

        return count;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n =  nums.length; 
        int count = 0 ; 
        int pre =  0 ; 
        HashMap<Integer,Integer>mp  = new HashMap<>();  
        mp.put(0 , 1);

        for(int  i =  0 ;  i < n ; i++){
            pre += nums[i] ; 
            int x = pre - k ; 
            if(mp.containsKey(x)){
                count += mp.get(x);
            }

            mp.put(pre , mp.getOrDefault(pre , 0 ) +1);
        }

        return count ; 
    }
}